src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs

```diff
@@ -334,9 +335,12 @@ internal void PropageOptions(CacheEntry parent)
             // We do this regardless of it gets cached because the tokens are associated with the value we'll return.
             if (_expirationTokens != null)
             {
-                foreach (var expirationToken in _expirationTokens)
+                lock (_lock)
                 {
-                    parent.AddExpirationToken(expirationToken);
+                    foreach (var expirationToken in _expirationTokens)
+                    {
+                        parent.AddExpirationToken(expirationToken);
+                    }
```

Comment:
Can you make ExpirationTokenRegistrations private? It's used inside the _lock (elsewhere) and doesn't appear to be accessed externally.